### PR TITLE
[Spec Decode] Make `propose_draft_token_ids` non-blocking for lower TTFT

### PR DESCRIPTION
**Pull request by WoosukKwon**
Currently, using spec decoding increases TTFT from `target_model_prefill_time` to `target_model_prefill_time + draft_model_prefill_time`, because the first token is returned together with the draft token ids.

This PR avoids this fake dependency by restructuring the `step` method, so that the sampled tokens can be returned to the scheduler without waiting for the draft tokens.

This optimization reduces TTFT by 0-30%, for long prefill or high QPS cases, while TPOT and overall throughput remain unchanged.

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -1825,7 +1835,7 @@ def propose_draft_token_ids(
                 indices = torch.tensor(indices, device=self.device)
                 hidden_states = sample_hidden_states[indices]
 
-            spec_token_ids = self.drafter.propose(
+            draft_token_ids = self.drafter.propose(
                 target_hidden_states=hidden_states,
                 sampling_metadata=sampling_metadata,
             )
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:1843]: Fixed the type annotation in Medusa and added a TODO on the optimization. cc @skylee-01

**Comment by WoosukKwon** [comment]
cc @njhill @robertgshaw2-redhat

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (skylee-01) [reply vllm/v1/worker/gpu_model_runner.py:1843]: Thanks for the work.

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -1785,14 +1785,24 @@ def execute_model(
...
             sampled_token_ids=valid_sampled_token_ids,
-            spec_token_ids=spec_token_ids,
             logprobs=logprobs_lists,
             prompt_logprobs_dict=prompt_logprobs_dict,
             pooler_output=[],
             kv_connector_output=kv_connector_output,
             num_nans_in_logits=num_nans_in_logits,
         )
 
+    def get_draft_token_ids(self) -> Optional[DraftTokenIds]:
```
Comment (njhill) [inline vllm/v1/worker/gpu_model_runner.py]: suggest renaming since it's not read-only:
```suggestion
    def take_draft_token_ids(self) -> Optional[DraftTokenIds]:
```

File: vllm/v1/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -1896,8 +1906,7 @@ def propose_draft_token_ids(
                 common_attn_metadata=common_attn_metadata,
                 mm_embeds=mm_embeds,
             )
-            spec_token_ids = draft_token_ids.tolist()
-        return spec_token_ids
+        return draft_token_ids
```
Comment (njhill) [inline vllm/v1/worker/gpu_model_runner.py:1911]: Would it help to start a non-blocking gpu->cpu transfer here?

File: vllm/v1/executor/abstract.py [hunk 4]
```diff
@@ -88,6 +88,10 @@ def execute_model(
                                      args=(scheduler_output, ))
         return output[0]
 
+    def get_draft_token_ids(self) -> Optional[DraftTokenIds]:
```
Comment (njhill) [inline vllm/v1/executor/abstract.py]: Similarly could consider renaming all of these to `take_draft_token_ids`

**Review by njhill — APPROVED** [review]
Thanks @WoosukKwon, this looks great

File: vllm/v1/worker/gpu_model_runner.py (hunk 3 above)
Comment (njhill) [reply vllm/v1/worker/gpu_model_runner.py:1911]: (with an appropriate sync obviously in the get_draft_tokens method)

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py]: good idea. Fixed!

File: vllm/v1/executor/abstract.py (hunk 4 above)
Comment (WoosukKwon) [reply vllm/v1/executor/abstract.py]: good idea. fixed!

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:1843]: @skylee-01 Can you please write a PR that implements this optimization? Basically, we want the medusa proposer to return a GPU tensor of shape `[num_reqs, num_spec_tokens]` instead of `list[list[int]]`.

File: vllm/v1/worker/gpu_model_runner.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:1911]: Great idea. Let's do it in a followup PR though?

Comment (njhill) [reply vllm/v1/worker/gpu_model_runner.py:1911]: sounds good

**Review by njhill — APPROVED** [review]
